develop/2017-05-29-fgrehm-analysis-on-meals-based-on-presence-information.py

```diff
@@ -0,0 +1,467 @@
+presences.sort_values('date', ascending=False)
+print("Presence records:", len(presences))
+
+presences = presences.query('presence == "Present"')
+print("Presence records with presence confirmed:", len(presences))
+
+# Match presence with session and deputy info
+presences = pd.merge(presences, sessions, on=['date', 'session'])
+print("Presence records with session matched:", len(presences))
+
+presences = pd.merge(deputies, presences, on='congressperson_document')
+print("Presence records with deputies matched:", len(presences))
+
+
+# In[13]:
+
+presences['first_session_at'] = presences['started_at']
+presences['last_session_at'] = presences['started_at']
+presences['total_sessions'] = 0
+aggregations = {
+    'first_session_at': 'min',
+    'last_session_at': 'max',
+    'total_sessions': 'count',
+    'session': lambda x: "{%s}" % ', '.join(x.astype(str)),
+}
+presences = presences.groupby(['congressperson_id', 'date'], as_index=False).agg(aggregations)
+print("Confirmed presences grouped by date and congressperson:", len(presences))
+
+
+# In[17]:
+
+meals_outside_df = meals.query('company_state != "DF"')
+meals_outside_df_while_in_df = pd.merge(
+    left=presences, 
+    right=meals_outside_df, 
+    left_on=['congressperson_id', 'date'],
+    right_on=['congressperson_id', 'issue_date']
+)
+print("Meals outside DF on days that the congressperson was there:", len(meals_outside_df_while_in_df))
+
+meals_outside_df_while_in_df = meals_outside_df_while_in_df.query('company_name != "GOL LINHAS AEREAS S.A."')
+print("Meals excluding expenses on flight:", len(meals_outside_df_while_in_df))
+
+meals_outside_df_while_in_df = meals_outside_df_while_in_df[
+    ~meals_outside_df_while_in_df.company_name.str.contains('HOTEL')
+]
+# We keep hotels out of the equation for this analysis since it they usually
+# have meals made on multiple days. Not to say that the bill might dated from
+# the day that the congressperson was back.
+print("Meals excluding expenses on hotels:", len(meals_outside_df_while_in_df))
+
+
+# In[18]:
+
+def score(meal):
+    return score_by_time(meal) + score_by_ts(meal)
+
+def score_by_time(meal):
+    if meal.total_sessions > 1:
+        return meal.last_session_at.hour - meal.first_session_at.hour + (meal.total_sessions - 1)
+    else:
+        return 0
+    
+def score_by_ts(meal):
+    if meal.timestamps == '':
+        return 0
+    
+    occurences = 0
+    for ts in meal.timestamps:
+        if ts == None:
+            continue
+        if (meal.first_session_at.hour-1) < ts.tm_hour < (meal.last_session_at.hour+1):
+            occurences += 1
+    return occurences * 10
+    
+suspects = meals_outside_df_while_in_df.copy()
+suspects.timestamps.fillna('', inplace=True)
+suspects['score'] = suspects.apply(score, axis='columns')
+suspects = suspects.query('score > 0').sort_values('score', ascending=False)
+print("Suspicious reimbursements:", len(suspects))
```

Comment:
For some reason this line doesn't get printed in the notebook... Is it possible to do a `Kernel -> Restart & Run All`? I hope this will then properly print this message.